---
title: "Methylation Association Analysis using MethParquet"
output: 
  rmarkdown::html_vignette:
    toc: true
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{Methylation Association Analysis using MethParquet}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

This vignette shows how to use [MethParquet](https://github.com/ZWangTen/MethParquet) to conduct association tests on normalized methylation data, from simple linear to mixed models incorporating kinship matrix as random effect to account for genetic similarity among samples, inspired by [GENESIS](https://bioconductor.org/packages/3.19/bioc/html/GENESIS.html). We also provide a short example for a more flexible association analysis using `flex_ewas()`, which accepts a regression model created by users as input to test each CpG site one after another. 

## Data

Load required packages
```{r setup,message=FALSE}
library(MethParquet)
library(arrow)
library(tidyverse)
library(limma)
library(GENESIS)
```

### Create Parquet database and MethList object

[MethParquet](https://github.com/ZWangTen/MethParquet) takes a `MethList` object created by `create_methlist()`. This object must contain the connection to the parquet database for the processed methylation data (in beta or M values), along with sample and CpG annotation in the `data.frame` format. Below is the example code to create the parquet database using `write_dataset()` from [arrow](https://arrow.apache.org/docs/r/index.html), followed by generation of MethList.

```{r}
# Load example data
data(phenoData)
data(chrAnnotation)
data(MethData)
head(MethData,c(3L,5L))
```

Create Parquet database in `path` where the methylation data is partitioned by chromosome.

```{r}
wdir <- getwd()
path <- paste0(wdir,'/data/Parquet_Directory')
MethData %>% group_by(CHR) %>% arrow::write_dataset(path,format = "parquet")
```

While creating MethList, users can choose to include only relevant columns from CpG and sample annotation to optimize data. Here only CpG names and position in chromosome (`Name` and `MAPINFO`), choromosome number `CHR` as well as gene names `UCSC_RefGene_Name` are included in the `mlist`.

```{r}
names(chrAnnotation)[c(2,12:13,16)]
mlist <- create_methlist(db_path = path,cpg_col_db='CpG',subject_annot = phenoData,
subject_col_keep='all',cpgAnnot_col_keep=c(2,12:13,16),cpg_annot = chrAnnotation,
subject_id='sample_id',cpg_col_annot='Name', gene_col_name = 'UCSC_RefGene_Name')
names(mlist)
```

## Association analysis

[MethParquet](https://github.com/ZWangTen/MethParquet) provides various models and ways to perform association analysis on methylation data. In this vignette we will show both the built-in models, as well as the flexible implementation using external ones. To streamline the pipeline, output of these analysis can be readily visualized by constructing QQ or Manhattan plots. 

### Simple and robust linear regression

`lm_ewas_outcome()` and `rlm_ewas_outcome()` regresses methylation on phenotypic trait and covariates, while returning the estimate effect of the trait and test statistics. We further accelerate simple linear regression by solving it through matrix and algebra operations, in a way that the CpG sites are tested in chunks and only relevant parameters are computed. That said, for same CpG list, `rlm_ewas_outcome()` would take longer to finish as it performs a "site by site" test. 
Here we illustrate an example of simple linear regression, with the other assuming the same usage.

```{r}
ewas_lm <- lm_ewas_outcome(db_obj=mlist,trait='bmi',covariates_string = c('age','sex'),out_position='MAPINFO')
head(ewas_lm)
```

As aforementioned, by specifying the CpG position in chromosome (`MAPINFO`), the output can be easily channeld to construct visualization plots. 

```{r,message=FALSE}
library(qqman)
library(CMplot)
qq(ewas_lm$p_value,main = "Q-Q plot of linear model", xlim = c(0, 4), ylim = c(0, 6), pch = 18, col = "blue4")
CMplot(ewas_lm[,c(1,8,9,6)],type='p',plot.type='m',dpi=300,file.output = FALSE,
       verbose=TRUE,width=400,height=172,chr.labels.angle=45,main = 'Manhattan_linear')
```

### Mixed linear model

As inspired by [GENESIS](https://bioconductor.org/packages/3.19/bioc/html/GENESIS.html), there are two steps involved when fitting a mixed model to test the association:

* `NullModel()` fits the null model without the CpG variable, where phenotypic trait is treated as the outcome. In particular, random effect is specified by covariance matrices (`cov.mat`). 

* `ewas_meth_exposure()` then runs the association testing by adding each CpG to the null model and returns a data frame with coefficient estimate and test statistics. 

```{r,message=FALSE}
# Create a random kinship matrix
library(Matrix)
library(psych)
kinship <- Matrix(runif(10000,min=-1,max=1), 100)
kinship <- forceSymmetric(kinship)
diag(kinship) <- 1-0.01*runif(100,min=-1,max=1)
row.names(kinship) <- phenoData$sample_id
colnames(kinship) <- row.names(kinship)
# Make the matrix positive definite
kinship <- cor.smooth(as.matrix(kinship))
kinship[1:4,1:4]
```

Now fit the null model with `age and `sex` as fixed and kinship matrix as random effect.

```{r}
m.null=NullModel(db_obj=mlist,trait='bmi',covariates_string = c('age','sex'),cov.mat = kinship,method='mixed',family='gaussian')
names(m.null)
m.null$NullModel$model$formula
```

Run association tests with fitted null model on chromosome 1-5.

```{r}
chr <- as.character(seq(from=1,to=5))
mixed=ewas_meth_exposure(db_obj=mlist,m.null,select_sites = FALSE,select_chr = chr, out_position=c('MAPINFO','Gene'))
head(mixed)
```

Users have great flexibility in customizing visualization. For instance, with gene name listed in the output, one can plot the CpG sites with gene as label using ggplot2. The CpGs on the top would be those with lowest p-values.

```{r}
ggplot(mixed, aes(x=CHR, y=-log10(p_value))) +
  geom_point(aes(color=as.factor(CHR))) + 
  geom_text(label=mixed$Gene,size=2,vjust=2) + theme(legend.position="none")
```

### Flexible EWAS

`flex_ewas` tests the CpG-phenotype association one by one using the provided external model. For example, the following robust linear regression takes methylation as exposure to compute its association with a quantitative trait `height`.

```{r}
rlm_function <- function(x) {
  mod <- base::suppressWarnings(MASS::rlm(as.formula(paste('height', "~x+", 'age')), data = phenoData))
  return(summary(mod)$coefficients[2,])
}

flex_rlm <- flex_ewas(mlist,rlm_function,out_position = c('MAPINFO','Gene'))
head(flex_rlm)
```

