---
title: "Using MethParquet for Methylation Risk Score and Differentially Methylated Regions"
output: 
  rmarkdown::html_vignette:
    toc: true
    fig_width: 6
    fig_height: 4
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{Using MethParquet for Methylation Risk Score and Differentially Methylated Regions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

This vignette shows how to use [MethParquet](https://github.com/ZWangTen/MethParquet) to compute methylation risk score (MRS) and analyze differentially methylated regions (DMRs) based on normalized methylation data. To detect DMRs we incorporated the software [Aclust2.0](https://github.com/OluwayioseOA/Aclust2.0). 

## Data

Load required packages
```{r setup,message=FALSE}
library(MethParquet)
library(arrow)
library(tidyverse)
library(knitr)
```

### Create Parquet database and MethList object

[MethParquet](https://github.com/ZWangTen/MethParquet) takes a `MethList` object created by `create_methlist()`. This object must contain the connection to the Parquet database for the processed methylation data (in beta or M values), along with sample and CpG annotation in the `data.frame` format. Below is the example code to create the parquet database using `write_parquet_meth`, leveraging the package [arrow](https://arrow.apache.org/docs/r/index.html), followed by generation of MethList. 
Note that MethData is only loaded to demonstrate the data structure. During creation of the Parquet database, this methylation matrix is not loaded into local memory.

```{r}
# Load example data
data(phenoData)
data(chrAnnotation)
MethData <- read.csv(system.file("extdata/MethData.csv", package="MethParquet"))
kable(head(MethData,c(3L,5L)))
```

Create Parquet database in `path` where the methylation data is partitioned by chromosome.

```{r}
wdir <- getwd()
methpath <- paste0(wdir,'/inst/extdata/MethData.csv')
path <- paste0(wdir,'/Parquet_Directory')
write_parquet_meth(data_path=methpath,format='csv',group_by='CHR',parquet_path = path)
```

While creating MethList, users can choose to include only relevant columns from CpG and sample annotation to optimize data. Here only CpG names and position in chromosome (`Name` and `MAPINFO`), choromosome number `CHR` as well as gene names `UCSC_RefGene_Name` are included in the `mlist`.

```{r}
names(chrAnnotation)[c(2,12:13,16)]
mlist <- create_methlist(db_path = path,cpg_col_db='CpG',subject_annot = phenoData,
subject_col_keep='all',cpgAnnot_col_keep=c(2,12:13,16),cpg_annot = chrAnnotation,
subject_id='sample_id',cpg_col_annot='Name', gene_col_name = 'UCSC_RefGene_Name')
names(mlist)
```

## Methylation Risk score (MRS)

`dev_meth_score()` can be used to compute MRS as weighted sum of an individualâ€™s DNAm values and pre-calculated weights associated with the phenotype of interest. Then `test_mrs` can test the MRS as an exposure for its association with a chosen phenotype, whether continuous or categorical, by fitting a linear, logistic or multinomial regression model.

### Compute MRS

Here we illustrate an example of computing a MRS for body mass index (BMI), and test its association with BMI. Note that for interpretation purposes, we can choose to scale the MRS.

```{r}
data(ewas_bmi) # Load data for BMI associated CpGs

mrs_bmi <- dev_meth_score(db_obj=mlist,cpg_col='CpG',meth_effect_data = ewas_bmi,effect_col = 'Effect', NAs_to_zero = TRUE,scale_score = TRUE) 
kable(head(mrs_bmi))
```

### Test MRS with a phenotype

Results of `test_mrs` is a list containing fitted model and the coefficients for the MRS. As shown below, the MRS for BMI showed a significant association with BMI for this example dataset. 

```{r,message=FALSE}
test_mrsBMI=test_mrs(db_obj = F,phe_data = mlist$subject_annot, outcome = 'bmi',covariates = F,mrs = mrs_bmi) 
summary(test_mrsBMI[[1]])
kable(test_mrsBMI[[2]])
```

## Differentially Methylated Regions (DMRs)

The software [Aclust2.0](https://github.com/OluwayioseOA/Aclust2.0) requires methylation data in the format that the rows and columns are CpGs and samples IDs. Since it's a rather time consuming process for DMR analysis, we provided a small toy data for DMR analysis. 
Here we provide an example code to obtain the betas data using `cpg_extract`. 

```{r,message=FALSE}
# Example code to generate subset of methylation data

# Sampling 5000 CpGs from CpG Annotation file
# cl<-mlist$cpg_annot %>% group_by('CHR') %>% sample_n(5000)

# Extract these 5000 CpGs
# extr<-cpg_extract(mlist,site_names=cl$Name)

# Get methylation data
# betas<-extr$cpg_data %>% dplyr::select(-CHR) %>% remove_rownames() %>% column_to_rownames('CpG')

# Resulted data:
betas_DMR<-read.csv(paste0(wdir,'/inst/extdata/MethData_DMR.csv'),row.names=1)
kable(head(betas_DMR[,1:3]))
```

The following script are from the `aclust2.0.R` script of the Aclust2.0 software.

Step 1:Call functions & import betas -------------------------------------

Download the code from Github into a named directory and set as working directory.

```{r,eval=FALSE}
library(tidyverse)
library(readr)
require(geepack)
require(data.table)
library("biomaRt") 
library("ChIPpeakAnno")
library("GenomicRanges")
library(tibble)
require('knitr')
setwd(path_to_Aclust2.0-main)
source(paste0(path_to_Aclust2.0-main,'aclust2.0_utils.R'))
```

Step 2: Get Illumina manifests for MM285, 450K, EPICv1, EPICv2 ---------

```{r,eval=FALSE}
manifest <- get_manifest("EPICv1") #to pull EPICv1 manifest
```

Step 3: Clustering of CpGs ----------------------------------------------

```{r,warning=FALSE,eval=FALSE}
##This step creates a list that accommodates three objects 
list.out <- find_cluster_list(probe.vec = rownames(betas_DMR), 
                              betas = betas_DMR, 
                              manifest = manifest, 
                              minimum.cluster.size = 2)
clusters.list <- list.out$clusters.list #required for optional GEE model in step 4
annot.betas <- list.out$annot.betas #required chromosomal annotation in step 5
cpg.clusters <- list.out$cpg.clusters #Includes only the CpG clusters which can be passed directly as clus object for gene annotations (step 5)
```

Step 4: Run gee model ---------------------------------------------------

```{r,eval=FALSE}
## Define exposure, covariates (optional) and sample.id
pheno<-read.csv(paste0(wdir,'/inst/extdata/PhenoData_DMR.csv'),row.names=1)
sample.id  <- pheno$sample_id
exposure <- pheno$bmi #(can also take categorical exposure variables)
covariates<- pheno %>% dplyr::select(age, sex)
identical(colnames(betas_DMR), sample.id) ##must be in the same order

###Adjusting for covariates
cluster.gee <- GEE.clusters(betas = betas_DMR, clusters.list = list.out$clusters.list, 
                            exposure = exposure,covariates= covariates, 
                            id = colnames(betas_DMR), working.cor = "ex", 
                            sample.id = sample.id) %>% 
  mutate(exposure_padjusted = p.adjust(exposure_pvalue, "BH")) ##adjusting for false positives

```

Step 5: Perform chromosomal annotations ---------------------------------

```{r,eval=FALSE}
#For cluster outputs from gee model in step 4
annot.betas <- list.out$annot.betas
clus <- cluster.gee ## (from step 4)
annotated.genes <- annot.clus.gene(annot.betas = annot.betas, 
                                   clus = clus, model = "hsa") 
```

Obtained DMR results by genes:

```{r}
annotated.genes <-read.csv(paste0(wdir,'/inst/extdata/DMR_results.csv'))
kable(annotated.genes)
unlink(path,recursive=TRUE)
```

