<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Flexible implementation of association functions on methylation data — flex_ewas • MethParquet</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Flexible implementation of association functions on methylation data — flex_ewas"><meta property="og:description" content="`flex_ewas` takes another regression function as input to test association between
each CpG site and the phenotype defined in the regression model"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethParquet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Assoc-test.html">Methylation Association Analysis using MethParquet</a>
    </li>
    <li>
      <a href="../articles/MRS_DMR.html">Using MethParquet for Methylation Risk Score and Differentially Methylated Regions</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Flexible implementation of association functions on methylation data</h1>
    
    <div class="hidden name"><code>flex_ewas.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>`flex_ewas` takes another regression function as input to test association between
each CpG site and the phenotype defined in the regression model</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">flex_ewas</span><span class="op">(</span></span>
<span>  <span class="va">db_obj</span>,</span>
<span>  <span class="va">fun</span>,</span>
<span>  phen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  select_sites <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  select_chr <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  gene_list <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  out_position <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  block_size <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>  NAs_to_zero <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>db_obj</dt>
<dd><p>MethList created using <code><a href="create_methlist.html">create_methlist</a></code>.</p></dd>


<dt>fun</dt>
<dd><p>A regression function applied to each CpG site</p></dd>


<dt>phen</dt>
<dd><p>Phenotype data based on which the function (`fun`) is built. Default: `NULL` is the subject annotation from the `db_obj`</p></dd>


<dt>select_sites</dt>
<dd><p>Character string consisting of CpG(s) to be tested. Default: `full` (all CpGs).</p></dd>


<dt>select_chr</dt>
<dd><p>Character string of chromosome number to select CpGs from. Default: `FALSE`.</p></dd>


<dt>gene_list</dt>
<dd><p>Character string of gene names to select CpGs from. Default: `FALSE`.</p></dd>


<dt>gene_col</dt>
<dd><p>Column for gene names in the CpG annotation data from `Methlist`.</p></dd>


<dt>out_position</dt>
<dd><p>Column names for CpG position in chromosome from `cpg_annot` of `MethList`.
If given, output will contain these columns and chromosome number.</p></dd>


<dt>block_size</dt>
<dd><p>Integer to specify number of CpGs in each iteration block. Default: 50,000.</p></dd>


<dt>NAs_to_zero</dt>
<dd><p>Convert missing values to 0 in methylation data. Default: `FALSE`.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data frame with results obtained using `fun`.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">phenoData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chrAnnotation</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">wdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">methpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,<span class="st">'MethData.csv'</span>,package<span class="op">=</span><span class="st">'MethParquet'</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create Parquet data and MethList</span></span></span>
<span class="r-in"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/Parquet_Directory'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="write_parquet_meth.html">write_parquet_meth</a></span><span class="op">(</span>data_path<span class="op">=</span><span class="va">methpath</span>,format<span class="op">=</span><span class="st">'csv'</span>,group_by<span class="op">=</span><span class="st">'CHR'</span>,parquet_path <span class="op">=</span> <span class="va">path</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">mlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_methlist.html">create_methlist</a></span><span class="op">(</span>db_path <span class="op">=</span> <span class="va">path</span>,cpg_col_db<span class="op">=</span><span class="st">'CpG'</span>,subject_annot <span class="op">=</span> <span class="va">phenoData</span>,</span></span>
<span class="r-in"><span>subject_col_keep<span class="op">=</span><span class="st">'all'</span>,cpgAnnot_col_keep<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fl">12</span><span class="op">:</span><span class="fl">13</span>,<span class="fl">16</span><span class="op">)</span>,cpg_annot <span class="op">=</span> <span class="va">chrAnnotation</span>,</span></span>
<span class="r-in"><span>subject_id<span class="op">=</span><span class="st">'sample_id'</span>,cpg_col_annot<span class="op">=</span><span class="st">'Name'</span>, gene_col_name <span class="op">=</span> <span class="st">'UCSC_RefGene_Name'</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create a robust linear regression function with methylation as exposure</span></span></span>
<span class="r-in"><span><span class="va">rlm_function</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/rlm.html" class="external-link">rlm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">'height'</span>, <span class="st">"~x+"</span>, <span class="st">'age'</span><span class="op">)</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">phenoData</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run flex_ewas</span></span></span>
<span class="r-in"><span><span class="va">flex_rlm</span> <span class="op">&lt;-</span> <span class="fu">flex_ewas</span><span class="op">(</span><span class="va">mlist</span>,<span class="va">rlm_function</span>,out_position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MAPINFO'</span>,<span class="st">'Gene'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in rlm.default(x, y, weights, method = method, wt.method = wt.method,     ...):</span> 'x' is singular: singular fits are not implemented in 'rlm'</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">flex_rlm</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'flex_rlm' not found</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">path</span>,recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ziqing Wang, Michael Cassidy, Danielle Clarkson-Townsend, Tamar Sofer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

