<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Methylation Association Analysis using MethParquet • MethParquet</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Methylation Association Analysis using MethParquet">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethParquet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Assoc-test.html">Methylation Association Analysis using MethParquet</a>
    </li>
    <li>
      <a href="../articles/MRS_DMR.html">Using MethParquet for Methylation Risk Score and Differentially Methylated Regions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Methylation Association Analysis using
MethParquet</h1>
            
            <h4 data-toc-skip class="date">2024-07-17</h4>
      

      <div class="hidden name"><code>Assoc-test.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette shows how to use <a href="https://github.com/ZWangTen/MethParquet" class="external-link">MethParquet</a> to
conduct association tests on normalized methylation data, from simple
linear to mixed models incorporating kinship matrix as random effect to
account for genetic similarity among samples, inspired by <a href="https://bioconductor.org/packages/3.19/bioc/html/GENESIS.html" class="external-link">GENESIS</a>.
We also provide a short example for a more flexible association analysis
using <code><a href="../reference/flex_ewas.html">flex_ewas()</a></code>, which accepts a regression model created
by users as input to test each CpG site one after another.</p>
<p>*Note that all results were based on toy datasets and hence are not
meant for significant biological interpretation.</p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>Load required packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MethParquet</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'arrow::string' by 'rlang::string' when</span></span>
<span><span class="co">#&gt; loading 'MethParquet'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/UW-GAC/GENESIS" class="external-link">GENESIS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="create-parquet-database-and-methlist-object">Create Parquet database and MethList object<a class="anchor" aria-label="anchor" href="#create-parquet-database-and-methlist-object"></a>
</h3>
<p><a href="https://github.com/ZWangTen/MethParquet" class="external-link">MethParquet</a>
takes a <code>MethList</code> object created by
<code><a href="../reference/create_methlist.html">create_methlist()</a></code>. This object must contain the connection
to the Parquet database for the processed methylation data (in beta or M
values), along with sample and CpG annotation in the
<code>data.frame</code> format. Below is the example code to create the
parquet database using <code>write_parquet_meth</code>, leveraging the
package <a href="https://arrow.apache.org/docs/r/index.html" class="external-link">arrow</a>,
followed by generation of MethList. Note that MethData is only loaded to
demonstrate the data structure. During creation of the Parquet database,
this methylation matrix is not loaded into local memory.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">phenoData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chrAnnotation</span><span class="op">)</span></span>
<span><span class="va">MethData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/MethData.csv"</span>, package<span class="op">=</span><span class="st">"MethParquet"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">MethData</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3L</span>,<span class="fl">5L</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CpG</th>
<th align="right">GSM1505103</th>
<th align="right">GSM1505113</th>
<th align="right">GSM1505099</th>
<th align="right">GSM1505117</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cg00567854</td>
<td align="right">0.655</td>
<td align="right">0.639</td>
<td align="right">0.745</td>
<td align="right">0.619</td>
</tr>
<tr class="even">
<td align="left">cg00714874</td>
<td align="right">0.683</td>
<td align="right">0.715</td>
<td align="right">0.616</td>
<td align="right">0.604</td>
</tr>
<tr class="odd">
<td align="left">cg00826902</td>
<td align="right">NA</td>
<td align="right">0.930</td>
<td align="right">NA</td>
<td align="right">0.958</td>
</tr>
</tbody>
</table>
<p>Create Parquet database in <code>path</code> where the methylation
data is partitioned by chromosome.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">methpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/inst/extdata/MethData.csv'</span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/Parquet_Directory'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_parquet_meth.html">write_parquet_meth</a></span><span class="op">(</span>data_path<span class="op">=</span><span class="va">methpath</span>,format<span class="op">=</span><span class="st">'csv'</span>,group_by<span class="op">=</span><span class="st">'CHR'</span>,parquet_path <span class="op">=</span> <span class="va">path</span><span class="op">)</span></span></code></pre></div>
<p>While creating MethList, users can choose to include only relevant
columns from CpG and sample annotation to optimize data. Here only CpG
names and position in chromosome (<code>Name</code> and
<code>MAPINFO</code>), choromosome number <code>CHR</code> as well as
gene names <code>UCSC_RefGene_Name</code> are included in the
<code>mlist</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">chrAnnotation</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">12</span><span class="op">:</span><span class="fl">13</span>,<span class="fl">16</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Name"              "CHR"               "MAPINFO"          </span></span>
<span><span class="co">#&gt; [4] "UCSC_RefGene_Name"</span></span>
<span><span class="va">mlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_methlist.html">create_methlist</a></span><span class="op">(</span>db_path <span class="op">=</span> <span class="va">path</span>,cpg_col_db<span class="op">=</span><span class="st">'CpG'</span>,subject_annot <span class="op">=</span> <span class="va">phenoData</span>,</span>
<span>subject_col_keep<span class="op">=</span><span class="st">'all'</span>,cpgAnnot_col_keep<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">12</span><span class="op">:</span><span class="fl">13</span>,<span class="fl">16</span><span class="op">)</span>,cpg_annot <span class="op">=</span> <span class="va">chrAnnotation</span>,</span>
<span>subject_id<span class="op">=</span><span class="st">'sample_id'</span>,cpg_col_annot<span class="op">=</span><span class="st">'Name'</span>, gene_col_name <span class="op">=</span> <span class="st">'UCSC_RefGene_Name'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mlist</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "db"            "subject_annot" "cpg_annot"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="association-analysis">Association analysis<a class="anchor" aria-label="anchor" href="#association-analysis"></a>
</h2>
<p><a href="https://github.com/ZWangTen/MethParquet" class="external-link">MethParquet</a>
provides various models and ways to perform association analysis on
methylation data. In this vignette we will show both the built-in
models, as well as the flexible implementation using external ones. To
streamline the pipeline, output of these analysis can be readily
visualized by constructing QQ or Manhattan plots.</p>
<div class="section level3">
<h3 id="simple-and-robust-linear-regression">Simple and robust linear regression<a class="anchor" aria-label="anchor" href="#simple-and-robust-linear-regression"></a>
</h3>
<p><code><a href="../reference/lm_ewas_outcome.html">lm_ewas_outcome()</a></code> and <code><a href="../reference/rlm_ewas_outcome.html">rlm_ewas_outcome()</a></code>
regresses methylation on phenotypic trait and covariates, while
returning the estimate effect of the trait and test statistics. We
further accelerate simple linear regression by solving it through matrix
and algebra operations, in a way that the CpG sites are tested in chunks
and only relevant parameters are computed. That said, for same CpG list,
<code><a href="../reference/rlm_ewas_outcome.html">rlm_ewas_outcome()</a></code> would take longer to finish as it
performs a “site by site” test. Here we illustrate an example of simple
linear regression, with the other assuming the same usage.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ewas_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lm_ewas_outcome.html">lm_ewas_outcome</a></span><span class="op">(</span>db_obj<span class="op">=</span><span class="va">mlist</span>,trait<span class="op">=</span><span class="st">'age'</span>,covariates_string <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'bmi'</span>,<span class="st">'sex'</span><span class="op">)</span>,out_position<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'CHR'</span>,<span class="st">'MAPINFO'</span><span class="op">)</span>,NAs_to_zero <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ewas_lm</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="12%">
<col width="12%">
<col width="11%">
<col width="12%">
<col width="11%">
<col width="11%">
<col width="11%">
<col width="4%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left">CpG</th>
<th align="right">estimate</th>
<th align="right">se</th>
<th align="right">t_stat</th>
<th align="right">t_stat_df</th>
<th align="right">p_value</th>
<th align="right">fdr_bh</th>
<th align="left">CHR</th>
<th align="right">MAPINFO</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cg00050692</td>
<td align="right">-0.0021709</td>
<td align="right">0.0007096</td>
<td align="right">-3.0590739</td>
<td align="right">96</td>
<td align="right">0.0028777</td>
<td align="right">0.0248571</td>
<td align="left">2</td>
<td align="right">25524877</td>
</tr>
<tr class="even">
<td align="left">cg00094412</td>
<td align="right">0.0002246</td>
<td align="right">0.0006355</td>
<td align="right">0.3534218</td>
<td align="right">96</td>
<td align="right">0.7245472</td>
<td align="right">0.8312166</td>
<td align="left">6</td>
<td align="right">29592854</td>
</tr>
<tr class="odd">
<td align="left">cg00108715</td>
<td align="right">-0.0004203</td>
<td align="right">0.0005670</td>
<td align="right">-0.7412196</td>
<td align="right">96</td>
<td align="right">0.4603696</td>
<td align="right">0.6295783</td>
<td align="left">3</td>
<td align="right">52565015</td>
</tr>
<tr class="even">
<td align="left">cg00138407</td>
<td align="right">0.0001527</td>
<td align="right">0.0010465</td>
<td align="right">0.1459213</td>
<td align="right">96</td>
<td align="right">0.8842895</td>
<td align="right">0.9340449</td>
<td align="left">3</td>
<td align="right">47386505</td>
</tr>
<tr class="odd">
<td align="left">cg00184953</td>
<td align="right">0.0021138</td>
<td align="right">0.0019015</td>
<td align="right">1.1116253</td>
<td align="right">96</td>
<td align="right">0.2690755</td>
<td align="right">0.4445127</td>
<td align="left">6</td>
<td align="right">31146222</td>
</tr>
<tr class="even">
<td align="left">cg00222799</td>
<td align="right">-0.0001143</td>
<td align="right">0.0005759</td>
<td align="right">-0.1984934</td>
<td align="right">96</td>
<td align="right">0.8430788</td>
<td align="right">0.9144712</td>
<td align="left">21</td>
<td align="right">43655464</td>
</tr>
</tbody>
</table>
<p>As aforementioned, by specifying the CpG position in chromosome
(<code>MAPINFO</code>), the output can be easily channeld to construct
visualization plots.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenturner/qqman" class="external-link">qqman</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/YinLiLin/CMplot" class="external-link">CMplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/qqman/man/qq.html" class="external-link">qq</a></span><span class="op">(</span><span class="va">ewas_lm</span><span class="op">$</span><span class="va">p_value</span>,main <span class="op">=</span> <span class="st">"Q-Q plot of linear model"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">18</span>, col <span class="op">=</span> <span class="st">"blue4"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Assoc-test_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/CMplot/man/CMplot-package.html" class="external-link">CMplot</a></span><span class="op">(</span><span class="va">ewas_lm</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span>,<span class="fl">9</span>,<span class="fl">6</span><span class="op">)</span><span class="op">]</span>,type<span class="op">=</span><span class="st">'p'</span>,plot.type<span class="op">=</span><span class="st">'m'</span>,dpi<span class="op">=</span><span class="fl">300</span>,file.output <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>       verbose<span class="op">=</span><span class="cn">TRUE</span>,width<span class="op">=</span><span class="fl">400</span>,height<span class="op">=</span><span class="fl">172</span>,chr.labels.angle<span class="op">=</span><span class="fl">45</span>,main <span class="op">=</span> <span class="st">'Manhattan_linear'</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Rectangular Manhattan plotting p_value.</span></span></code></pre></div>
<p><img src="Assoc-test_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="mixed-linear-model">Mixed linear model<a class="anchor" aria-label="anchor" href="#mixed-linear-model"></a>
</h3>
<p>As inspired by <a href="https://bioconductor.org/packages/3.19/bioc/html/GENESIS.html" class="external-link">GENESIS</a>,
there are two steps involved when fitting a mixed model to test the
association:</p>
<ul>
<li><p><code><a href="../reference/NullModel.html">NullModel()</a></code> fits the null model without the CpG
variable, where phenotypic trait is treated as the outcome. In
particular, random effect is specified by covariance matrices
(<code>cov.mat</code>).</p></li>
<li><p><code><a href="../reference/ewas_meth_exposure.html">ewas_meth_exposure()</a></code> then runs the association
testing by adding each CpG to the null model and returns a data frame
with coefficient estimate and test statistics.</p></li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a random kinship matrix</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://personality-project.org/r/psych/" class="external-link">psych</a></span><span class="op">)</span></span>
<span><span class="va">kinship</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>,min<span class="op">=</span><span class="op">-</span><span class="fl">1</span>,max<span class="op">=</span><span class="fl">1</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">kinship</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/forceSymmetric-methods.html" class="external-link">forceSymmetric</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">-</span><span class="fl">0.01</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>,min<span class="op">=</span><span class="op">-</span><span class="fl">1</span>,max<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">phenoData</span><span class="op">$</span><span class="va">sample_id</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span></span>
<span><span class="co"># Make the matrix positive definite</span></span>
<span><span class="va">kinship</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/psych/man/cor.smooth.html" class="external-link">cor.smooth</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in cor.smooth(as.matrix(kinship)): Matrix was not positive definite,</span></span>
<span><span class="co">#&gt; smoothing was done</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">GSM1505103</th>
<th align="right">GSM1505113</th>
<th align="right">GSM1505099</th>
<th align="right">GSM1505117</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">GSM1505103</td>
<td align="right">1.0000000</td>
<td align="right">0.0979275</td>
<td align="right">-0.0949970</td>
<td align="right">0.1499061</td>
</tr>
<tr class="even">
<td align="left">GSM1505113</td>
<td align="right">0.0979275</td>
<td align="right">1.0000000</td>
<td align="right">0.2204344</td>
<td align="right">-0.0939170</td>
</tr>
<tr class="odd">
<td align="left">GSM1505099</td>
<td align="right">-0.0949970</td>
<td align="right">0.2204344</td>
<td align="right">1.0000000</td>
<td align="right">-0.0413349</td>
</tr>
<tr class="even">
<td align="left">GSM1505117</td>
<td align="right">0.1499061</td>
<td align="right">-0.0939170</td>
<td align="right">-0.0413349</td>
<td align="right">1.0000000</td>
</tr>
</tbody>
</table>
<p>Now fit the null model with <code>age and</code>sex` as fixed and
kinship matrix as random effect.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m.null</span><span class="op">=</span><span class="fu"><a href="../reference/NullModel.html">NullModel</a></span><span class="op">(</span>db_obj<span class="op">=</span><span class="va">mlist</span>,trait<span class="op">=</span><span class="st">'bmi'</span>,covariates_string <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'age'</span>,<span class="st">'sex'</span><span class="op">)</span>,cov.mat <span class="op">=</span> <span class="va">kinship</span>,method<span class="op">=</span><span class="st">'mixed'</span>,family<span class="op">=</span><span class="st">'gaussian'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing Variance Component Estimates...</span></span>
<span><span class="co">#&gt; Sigma^2_A     log-lik     RSS</span></span>
<span><span class="co">#&gt; [1]   71.5596274   71.5596274 -378.5245838    0.9811488</span></span>
<span><span class="co">#&gt; [1]   27.249861   81.587682 -374.919423    1.058672</span></span>
<span><span class="co">#&gt; [1]   16.540708   93.491124 -374.620423    1.014747</span></span>
<span><span class="co">#&gt; [1]   16.998809   94.672170 -374.620314    1.000216</span></span>
<span><span class="co">#&gt; [1]   16.99712   94.69694 -374.62031    1.00000</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m.null</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "NullModel" "Y"         "X"</span></span>
<span><span class="va">m.null</span><span class="op">$</span><span class="va">NullModel</span><span class="op">$</span><span class="va">model</span><span class="op">$</span><span class="va">formula</span></span>
<span><span class="co">#&gt; [1] "bmi ~ age + sex + (1|A)"</span></span></code></pre></div>
<p>Run association tests with fitted null model on chromosome 1-5.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">1</span>,to<span class="op">=</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mixed</span><span class="op">=</span><span class="fu"><a href="../reference/ewas_meth_exposure.html">ewas_meth_exposure</a></span><span class="op">(</span>db_obj<span class="op">=</span><span class="va">mlist</span>,<span class="va">m.null</span>,select_sites <span class="op">=</span> <span class="cn">FALSE</span>,select_chr <span class="op">=</span> <span class="va">chr</span>, out_position<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'CHR'</span>,<span class="st">'MAPINFO'</span>,<span class="st">'Gene'</span><span class="op">)</span>,NAs_to_zero <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mixed</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="11%">
<col width="11%">
<col width="11%">
<col width="11%">
<col width="10%">
<col width="10%">
<col width="4%">
<col width="10%">
<col width="21%">
</colgroup>
<thead><tr class="header">
<th align="left">CpG</th>
<th align="right">Estimate</th>
<th align="right">Score</th>
<th align="right">Score.Stat</th>
<th align="right">p_value</th>
<th align="right">fdr_bh</th>
<th align="left">CHR</th>
<th align="right">MAPINFO</th>
<th align="left">Gene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cg00050692</td>
<td align="right">-7.793037</td>
<td align="right">-0.0886922</td>
<td align="right">-0.8313733</td>
<td align="right">0.4057628</td>
<td align="right">0.4780769</td>
<td align="left">2</td>
<td align="right">25524877</td>
<td align="left">DNMT3A;DNMT3A;DNMT3A</td>
</tr>
<tr class="even">
<td align="left">cg00108715</td>
<td align="right">16.977956</td>
<td align="right">0.1273396</td>
<td align="right">1.4703625</td>
<td align="right">0.1414636</td>
<td align="right">0.2215022</td>
<td align="left">3</td>
<td align="right">52565015</td>
<td align="left">NT5DC2;NT5DC2</td>
</tr>
<tr class="odd">
<td align="left">cg00138407</td>
<td align="right">-9.687622</td>
<td align="right">-0.2583517</td>
<td align="right">-1.5820282</td>
<td align="right">0.1136431</td>
<td align="right">0.1904723</td>
<td align="left">3</td>
<td align="right">47386505</td>
<td align="left">KLHL18</td>
</tr>
<tr class="even">
<td align="left">cg00567854</td>
<td align="right">-20.477495</td>
<td align="right">-0.0892299</td>
<td align="right">-1.3517416</td>
<td align="right">0.1764580</td>
<td align="right">0.2458326</td>
<td align="left">1</td>
<td align="right">203273693</td>
<td align="left">BTG2</td>
</tr>
<tr class="odd">
<td align="left">cg00714874</td>
<td align="right">-2.182466</td>
<td align="right">-0.0117571</td>
<td align="right">-0.1601856</td>
<td align="right">0.8727349</td>
<td align="right">0.9185631</td>
<td align="left">1</td>
<td align="right">161679461</td>
<td align="left">FCRLA</td>
</tr>
<tr class="even">
<td align="left">cg00741986</td>
<td align="right">-44.457157</td>
<td align="right">-0.1473435</td>
<td align="right">-2.5593894</td>
<td align="right">0.0104856</td>
<td align="right">0.0258522</td>
<td align="left">4</td>
<td align="right">2748332</td>
<td align="left">TNIP2;TNIP2</td>
</tr>
</tbody>
</table>
<p>Users have great flexibility in customizing visualization. For
instance, with gene name listed in the output, one can plot the CpG
sites with gene as label using ggplot2. The CpGs on the top would be
those with lowest p-values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mixed</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">CHR</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">p_value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">CHR</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">mixed</span><span class="op">$</span><span class="va">Gene</span>,size<span class="op">=</span><span class="fl">2</span>,vjust<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Assoc-test_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="flexible-ewas">Flexible EWAS<a class="anchor" aria-label="anchor" href="#flexible-ewas"></a>
</h3>
<p><code>flex_ewas</code> tests the CpG-phenotype association one by one
using the provided external model. For example, the following robust
linear regression takes methylation as exposure to compute its
association with a quantitative trait <code>height</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rlm_function</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/rlm.html" class="external-link">rlm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">as.formula</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">'age'</span>, <span class="st">"~x+"</span>, <span class="st">'age'</span><span class="op">)</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">phenoData</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">flex_rlm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flex_ewas.html">flex_ewas</a></span><span class="op">(</span><span class="va">mlist</span>,<span class="va">rlm_function</span>,out_position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MAPINFO'</span>,<span class="st">'Gene'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">flex_rlm</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="11%">
<col width="11%">
<col width="11%">
<col width="11%">
<col width="4%">
<col width="9%">
<col width="38%">
</colgroup>
<thead><tr class="header">
<th align="left">CpG</th>
<th align="right">Value</th>
<th align="right">Std. Error</th>
<th align="right">t value</th>
<th align="left">CHR</th>
<th align="right">MAPINFO</th>
<th align="left">Gene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cg00050692</td>
<td align="right">-42.581828</td>
<td align="right">14.62981</td>
<td align="right">-2.9106215</td>
<td align="left">2</td>
<td align="right">25524877</td>
<td align="left">DNMT3A;DNMT3A;DNMT3A</td>
</tr>
<tr class="even">
<td align="left">cg00094412</td>
<td align="right">9.641751</td>
<td align="right">20.83006</td>
<td align="right">0.4628767</td>
<td align="left">6</td>
<td align="right">29592854</td>
<td align="left">GABBR1;GABBR1;GABBR1</td>
</tr>
<tr class="odd">
<td align="left">cg00108715</td>
<td align="right">-12.901337</td>
<td align="right">21.28209</td>
<td align="right">-0.6062063</td>
<td align="left">3</td>
<td align="right">52565015</td>
<td align="left">NT5DC2;NT5DC2</td>
</tr>
<tr class="even">
<td align="left">cg00138407</td>
<td align="right">13.860760</td>
<td align="right">36.19346</td>
<td align="right">0.3829631</td>
<td align="left">3</td>
<td align="right">47386505</td>
<td align="left">KLHL18</td>
</tr>
<tr class="odd">
<td align="left">cg00184953</td>
<td align="right">-29.013044</td>
<td align="right">40.09269</td>
<td align="right">-0.7236491</td>
<td align="left">6</td>
<td align="right">31146222</td>
<td align="left">PSORS1C3</td>
</tr>
<tr class="even">
<td align="left">cg00222799</td>
<td align="right">-2.777217</td>
<td align="right">21.06250</td>
<td align="right">-0.1318560</td>
<td align="left">21</td>
<td align="right">43655464</td>
<td align="left">ABCG1;ABCG1;ABCG1;ABCG1;ABCG1;ABCG1</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">path</span>,recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ziqing Wang, Michael Cassidy, Danielle Clarkson-Townsend, Tamar Sofer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
