<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using MethParquet for Methylation Risk Score and Differentially Methylated Regions â€¢ MethParquet</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using MethParquet for Methylation Risk Score and Differentially Methylated Regions">
<meta property="og:description" content="MethParquet">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethParquet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Assoc-test.html">Methylation Association Analysis using MethParquet</a>
    </li>
    <li>
      <a href="../articles/MRS_DMR.html">Using MethParquet for Methylation Risk Score and Differentially Methylated Regions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using MethParquet for Methylation Risk Score and
Differentially Methylated Regions</h1>
            
            <h4 data-toc-skip class="date">2024-05-28</h4>
      
      
      <div class="hidden name"><code>MRS_DMR.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette shows how to use <a href="https://github.com/ZWangTen/MethParquet" class="external-link">MethParquet</a> to
compute methylation risk score (MRS) and analyze differentially
methylated regions (DMRs) based on normalized methylation data. To
detect DMRs we incorporated the software <a href="https://github.com/OluwayioseOA/Aclust2.0" class="external-link">Aclust2.0</a>.</p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>Load required packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MethParquet</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'arrow::string' by 'rlang::string' when</span></span>
<span><span class="co">#&gt; loading 'MethParquet'</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="create-parquet-database-and-methlist-object">Create Parquet database and MethList object<a class="anchor" aria-label="anchor" href="#create-parquet-database-and-methlist-object"></a>
</h3>
<p><a href="https://github.com/ZWangTen/MethParquet" class="external-link">MethParquet</a>
takes a <code>MethList</code> object created by
<code><a href="../reference/create_methlist.html">create_methlist()</a></code>. This object must contain the connection
to the Parquet database for the processed methylation data (in beta or M
values), along with sample and CpG annotation in the
<code>data.frame</code> format. Below is the example code to create the
parquet database using <code>write_parquet_meth</code>, leveraging the
package <a href="https://arrow.apache.org/docs/r/index.html" class="external-link">arrow</a>,
followed by generation of MethList. Note that MethData is only loaded to
demonstrate the data structure. During creation of the Parquet database,
this methylation matrix is not loaded into local memory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">phenoData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chrAnnotation</span><span class="op">)</span></span>
<span><span class="va">MethData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/MethData.csv"</span>, package<span class="op">=</span><span class="st">"MethParquet"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">MethData</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3L</span>,<span class="fl">5L</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CpG</th>
<th align="right">GSM1505103</th>
<th align="right">GSM1505113</th>
<th align="right">GSM1505099</th>
<th align="right">GSM1505117</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cg00567854</td>
<td align="right">0.655</td>
<td align="right">0.639</td>
<td align="right">0.745</td>
<td align="right">0.619</td>
</tr>
<tr class="even">
<td align="left">cg00714874</td>
<td align="right">0.683</td>
<td align="right">0.715</td>
<td align="right">0.616</td>
<td align="right">0.604</td>
</tr>
<tr class="odd">
<td align="left">cg00826902</td>
<td align="right">NA</td>
<td align="right">0.930</td>
<td align="right">NA</td>
<td align="right">0.958</td>
</tr>
</tbody>
</table>
<p>Create Parquet database in <code>path</code> where the methylation
data is partitioned by chromosome.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">methpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/inst/extdata/MethData.csv'</span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/Parquet_Directory'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_parquet_meth.html">write_parquet_meth</a></span><span class="op">(</span>data_path<span class="op">=</span><span class="va">methpath</span>,format<span class="op">=</span><span class="st">'csv'</span>,group_by<span class="op">=</span><span class="st">'CHR'</span>,parquet_path <span class="op">=</span> <span class="va">path</span><span class="op">)</span></span></code></pre></div>
<p>While creating MethList, users can choose to include only relevant
columns from CpG and sample annotation to optimize data. Here only CpG
names and position in chromosome (<code>Name</code> and
<code>MAPINFO</code>), choromosome number <code>CHR</code> as well as
gene names <code>UCSC_RefGene_Name</code> are included in the
<code>mlist</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">chrAnnotation</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">12</span><span class="op">:</span><span class="fl">13</span>,<span class="fl">16</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Name"              "CHR"               "MAPINFO"          </span></span>
<span><span class="co">#&gt; [4] "UCSC_RefGene_Name"</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_methlist.html">create_methlist</a></span><span class="op">(</span>db_path <span class="op">=</span> <span class="va">path</span>,cpg_col_db<span class="op">=</span><span class="st">'CpG'</span>,subject_annot <span class="op">=</span> <span class="va">phenoData</span>,</span>
<span>subject_col_keep<span class="op">=</span><span class="st">'all'</span>,cpgAnnot_col_keep<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">12</span><span class="op">:</span><span class="fl">13</span>,<span class="fl">16</span><span class="op">)</span>,cpg_annot <span class="op">=</span> <span class="va">chrAnnotation</span>,</span>
<span>subject_id<span class="op">=</span><span class="st">'sample_id'</span>,cpg_col_annot<span class="op">=</span><span class="st">'Name'</span>, gene_col_name <span class="op">=</span> <span class="st">'UCSC_RefGene_Name'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">mlist</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "db"            "subject_annot" "cpg_annot"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="methylation-risk-score-mrs">Methylation Risk score (MRS)<a class="anchor" aria-label="anchor" href="#methylation-risk-score-mrs"></a>
</h2>
<p><code><a href="../reference/dev_meth_score.html">dev_meth_score()</a></code> can be used to compute MRS as weighted
sum of an individualâ€™s DNAm values and pre-calculated weights associated
with the phenotype of interest. Then <code>test_mrs</code> can test the
MRS as an exposure for its association with a chosen phenotype, whether
continuous or categorical, by fitting a linear, logistic or multinomial
regression model.</p>
<div class="section level3">
<h3 id="compute-mrs">Compute MRS<a class="anchor" aria-label="anchor" href="#compute-mrs"></a>
</h3>
<p>Here we illustrate an example of computing a MRS for body mass index
(BMI), and test its association with BMI. Note that for interpretation
purposes, we can choose to scale the MRS.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ewas_bmi</span><span class="op">)</span> <span class="co"># Load data for BMI associated CpGs</span></span>
<span></span>
<span><span class="va">mrs_bmi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dev_meth_score.html">dev_meth_score</a></span><span class="op">(</span>db_obj<span class="op">=</span><span class="va">mlist</span>,cpg_col<span class="op">=</span><span class="st">'CpG'</span>,meth_effect_data <span class="op">=</span> <span class="va">ewas_bmi</span>,effect_col <span class="op">=</span> <span class="st">'Effect'</span>, NAs_to_zero <span class="op">=</span> <span class="cn">TRUE</span>,scale_score <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mrs_bmi</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table"><tbody>
<tr class="odd">
<td align="left">GSM1505103</td>
<td align="right">0.1578201</td>
</tr>
<tr class="even">
<td align="left">GSM1505113</td>
<td align="right">-1.2952102</td>
</tr>
<tr class="odd">
<td align="left">GSM1505099</td>
<td align="right">0.1640210</td>
</tr>
<tr class="even">
<td align="left">GSM1505117</td>
<td align="right">-0.8833481</td>
</tr>
<tr class="odd">
<td align="left">GSM1505102</td>
<td align="right">-0.1880272</td>
</tr>
<tr class="even">
<td align="left">GSM1505101</td>
<td align="right">-1.3378684</td>
</tr>
</tbody></table>
</div>
<div class="section level3">
<h3 id="test-mrs-with-a-phenotype">Test MRS with a phenotype<a class="anchor" aria-label="anchor" href="#test-mrs-with-a-phenotype"></a>
</h3>
<p>Results of <code>test_mrs</code> is a list containing fitted model
and the coefficients for the MRS. As shown below, the MRS for BMI showed
a significant association with BMI for this example dataset.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_mrsBMI</span><span class="op">=</span><span class="fu"><a href="../reference/test_mrs.html">test_mrs</a></span><span class="op">(</span>db_obj <span class="op">=</span> <span class="cn">F</span>,phe_data <span class="op">=</span> <span class="va">mlist</span><span class="op">$</span><span class="va">subject_annot</span>, outcome <span class="op">=</span> <span class="st">'bmi'</span>,covariates <span class="op">=</span> <span class="cn">F</span>,mrs <span class="op">=</span> <span class="va">mrs_bmi</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">test_mrsBMI</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; lm(formula = y ~ x)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residuals:</span></span>
<span><span class="co">#&gt;     Min      1Q  Median      3Q     Max </span></span>
<span><span class="co">#&gt; -25.611  -6.842  -0.928   5.955  32.532 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients: (1 not defined because of singularities)</span></span>
<span><span class="co">#&gt;              Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">#&gt; (Intercept)    40.894      1.124  36.377  &lt; 2e-16 ***</span></span>
<span><span class="co">#&gt; x(Intercept)       NA         NA      NA       NA    </span></span>
<span><span class="co">#&gt; xmrs           -4.245      1.130  -3.757 0.000292 ***</span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residual standard error: 11.24 on 98 degrees of freedom</span></span>
<span><span class="co">#&gt; Multiple R-squared:  0.1259, Adjusted R-squared:  0.117 </span></span>
<span><span class="co">#&gt; F-statistic: 14.12 on 1 and 98 DF,  p-value: 0.000292</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">test_mrsBMI</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">mrs</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Estimate</td>
<td align="right">-4.244919</td>
</tr>
<tr class="even">
<td align="left">Std. Error</td>
<td align="right">1.129836</td>
</tr>
<tr class="odd">
<td align="left">t value</td>
<td align="right">-3.757111</td>
</tr>
<tr class="even">
<td align="left">Pr(&gt;|t|)</td>
<td align="right">0.000292</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="differentially-methylated-regions-dmrs">Differentially Methylated Regions (DMRs)<a class="anchor" aria-label="anchor" href="#differentially-methylated-regions-dmrs"></a>
</h2>
<p>The software <a href="https://github.com/OluwayioseOA/Aclust2.0" class="external-link">Aclust2.0</a> requires
methylation data in the format that the rows and columns are CpGs and
samples IDs. Since itâ€™s a rather time consuming process for DMR
analysis, we provided a small toy data for DMR analysis. Here we provide
an example code to obtain the betas data using
<code>cpg_extract</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example code to generate subset of methylation data</span></span>
<span></span>
<span><span class="co"># Sampling 5000 CpGs from CpG Annotation file</span></span>
<span><span class="co"># cl&lt;-mlist$cpg_annot %&gt;% group_by('CHR') %&gt;% sample_n(5000)</span></span>
<span></span>
<span><span class="co"># Extract these 5000 CpGs</span></span>
<span><span class="co"># extr&lt;-cpg_extract(mlist,site_names=cl$Name)</span></span>
<span></span>
<span><span class="co"># Get methylation data</span></span>
<span><span class="co"># betas&lt;-extr$cpg_data %&gt;% dplyr::select(-CHR) %&gt;% remove_rownames() %&gt;% column_to_rownames('CpG')</span></span>
<span></span>
<span><span class="co"># Resulted data:</span></span>
<span><span class="va">betas_DMR</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/inst/extdata/MethData_DMR.csv'</span><span class="op">)</span>,row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">betas_DMR</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">GSM1505103</th>
<th align="right">GSM1505113</th>
<th align="right">GSM1505099</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cg14657517</td>
<td align="right">0.034</td>
<td align="right">0.062</td>
<td align="right">0.043</td>
</tr>
<tr class="even">
<td align="left">cg09810313</td>
<td align="right">0.021</td>
<td align="right">0.021</td>
<td align="right">0.011</td>
</tr>
<tr class="odd">
<td align="left">cg27186774</td>
<td align="right">0.007</td>
<td align="right">0.017</td>
<td align="right">0.012</td>
</tr>
<tr class="even">
<td align="left">cg05592035</td>
<td align="right">0.416</td>
<td align="right">0.428</td>
<td align="right">0.477</td>
</tr>
<tr class="odd">
<td align="left">cg24775616</td>
<td align="right">0.153</td>
<td align="right">0.206</td>
<td align="right">0.047</td>
</tr>
<tr class="even">
<td align="left">cg21934337</td>
<td align="right">0.836</td>
<td align="right">0.930</td>
<td align="right">0.830</td>
</tr>
</tbody>
</table>
<p>The following script are from the <code>aclust2.0.R</code> script of
the Aclust2.0 software.</p>
<p>Step 1:Call functions &amp; import betas â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”-</p>
<p>Download the code from Github into a named directory and set as
working directory.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">geepack</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"ChIPpeakAnno"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/" class="external-link">'knitr'</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">path_to_Aclust2.0</span><span class="op">-</span><span class="va">main</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path_to_Aclust2.0</span><span class="op">-</span><span class="va">main</span>,<span class="st">'aclust2.0_utils.R'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Step 2: Get Illumina manifests for MM285, 450K, EPICv1, EPICv2
â€”â€”â€”</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">manifest</span> <span class="op">&lt;-</span> <span class="fu">get_manifest</span><span class="op">(</span><span class="st">"EPICv1"</span><span class="op">)</span> <span class="co">#to pull EPICv1 manifest</span></span></code></pre></div>
<p>Step 3: Clustering of CpGs â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”-</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##This step creates a list that accommodates three objects </span></span>
<span><span class="va">list.out</span> <span class="op">&lt;-</span> <span class="fu">find_cluster_list</span><span class="op">(</span>probe.vec <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">betas_DMR</span><span class="op">)</span>, </span>
<span>                              betas <span class="op">=</span> <span class="va">betas_DMR</span>, </span>
<span>                              manifest <span class="op">=</span> <span class="va">manifest</span>, </span>
<span>                              minimum.cluster.size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">clusters.list</span> <span class="op">&lt;-</span> <span class="va">list.out</span><span class="op">$</span><span class="va">clusters.list</span> <span class="co">#required for optional GEE model in step 4</span></span>
<span><span class="va">annot.betas</span> <span class="op">&lt;-</span> <span class="va">list.out</span><span class="op">$</span><span class="va">annot.betas</span> <span class="co">#required chromosomal annotation in step 5</span></span>
<span><span class="va">cpg.clusters</span> <span class="op">&lt;-</span> <span class="va">list.out</span><span class="op">$</span><span class="va">cpg.clusters</span> <span class="co">#Includes only the CpG clusters which can be passed directly as clus object for gene annotations (step 5)</span></span></code></pre></div>
<p>Step 4: Run gee model â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define exposure, covariates (optional) and sample.id</span></span>
<span><span class="va">pheno</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/inst/extdata/PhenoData_DMR.csv'</span><span class="op">)</span>,row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sample.id</span>  <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">$</span><span class="va">sample_id</span></span>
<span><span class="va">exposure</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">$</span><span class="va">bmi</span> <span class="co">#(can also take categorical exposure variables)</span></span>
<span><span class="va">covariates</span><span class="op">&lt;-</span> <span class="va">pheno</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">age</span>, <span class="va">sex</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">betas_DMR</span><span class="op">)</span>, <span class="va">sample.id</span><span class="op">)</span> <span class="co">##must be in the same order</span></span>
<span></span>
<span><span class="co">###Adjusting for covariates</span></span>
<span><span class="va">cluster.gee</span> <span class="op">&lt;-</span> <span class="fu">GEE.clusters</span><span class="op">(</span>betas <span class="op">=</span> <span class="va">betas_DMR</span>, clusters.list <span class="op">=</span> <span class="va">list.out</span><span class="op">$</span><span class="va">clusters.list</span>, </span>
<span>                            exposure <span class="op">=</span> <span class="va">exposure</span>,covariates<span class="op">=</span> <span class="va">covariates</span>, </span>
<span>                            id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">betas_DMR</span><span class="op">)</span>, working.cor <span class="op">=</span> <span class="st">"ex"</span>, </span>
<span>                            sample.id <span class="op">=</span> <span class="va">sample.id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>exposure_padjusted <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">exposure_pvalue</span>, <span class="st">"BH"</span><span class="op">)</span><span class="op">)</span> <span class="co">##adjusting for false positives</span></span></code></pre></div>
<p>Step 5: Perform chromosomal annotations â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”â€”</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#For cluster outputs from gee model in step 4</span></span>
<span><span class="va">annot.betas</span> <span class="op">&lt;-</span> <span class="va">list.out</span><span class="op">$</span><span class="va">annot.betas</span></span>
<span><span class="va">clus</span> <span class="op">&lt;-</span> <span class="va">cluster.gee</span> <span class="co">## (from step 4)</span></span>
<span><span class="va">annotated.genes</span> <span class="op">&lt;-</span> <span class="fu">annot.clus.gene</span><span class="op">(</span>annot.betas <span class="op">=</span> <span class="va">annot.betas</span>, </span>
<span>                                   clus <span class="op">=</span> <span class="va">clus</span>, model <span class="op">=</span> <span class="st">"hsa"</span><span class="op">)</span> </span></code></pre></div>
<p>Obtained DMR results by genes:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotated.genes</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">wdir</span>,<span class="st">'/inst/extdata/DMR_results.csv'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">annotated.genes</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="5%">
<col width="6%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="4%">
<col width="6%">
<col width="8%">
<col width="5%">
<col width="1%">
<col width="5%">
<col width="4%">
<col width="5%">
<col width="5%">
<col width="6%">
<col width="6%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left">feature</th>
<th align="left">external_gene_name</th>
<th align="left">seqnames</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="right">width</th>
<th align="left">strand</th>
<th align="right">probe</th>
<th align="left">cluster_name</th>
<th align="right">n_sites_in_cluster</th>
<th align="left">cluster_sites</th>
<th align="left">cluster_name.1</th>
<th align="right">peak</th>
<th align="right">start_position</th>
<th align="right">end_position</th>
<th align="left">feature_strand</th>
<th align="left">insideFeature</th>
<th align="right">distancetoFeature</th>
<th align="right">shortestDistance</th>
<th align="left">fromOverlappingOrNearest</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ENSE00001174442</td>
<td align="left">IL6R</td>
<td align="left">chr1</td>
<td align="right">154404902</td>
<td align="right">154404926</td>
<td align="right">25</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_1</td>
<td align="right">2</td>
<td align="left">;cg01435678;cg23609982</td>
<td align="left">cluster_1</td>
<td align="right">1</td>
<td align="right">154405193</td>
<td align="right">154405714</td>
<td align="left">+</td>
<td align="left">upstream</td>
<td align="right">-279</td>
<td align="right">267</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="even">
<td align="left">ENSE00001338726</td>
<td align="left">IRX6</td>
<td align="left">chr16</td>
<td align="right">55328429</td>
<td align="right">55328841</td>
<td align="right">413</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_5</td>
<td align="right">2</td>
<td align="left">;cg05099576;cg04834436</td>
<td align="left">cluster_5</td>
<td align="right">5</td>
<td align="right">55328700</td>
<td align="right">55329311</td>
<td align="left">+</td>
<td align="left">overlapStart</td>
<td align="right">-65</td>
<td align="right">141</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="odd">
<td align="left">ENSE00001403134</td>
<td align="left">ATP2A1</td>
<td align="left">chr16</td>
<td align="right">28878308</td>
<td align="right">28878778</td>
<td align="right">471</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_4</td>
<td align="right">2</td>
<td align="left">;cg20877811;cg00000292</td>
<td align="left">cluster_4</td>
<td align="right">4</td>
<td align="right">28878488</td>
<td align="right">28878789</td>
<td align="left">+</td>
<td align="left">overlapStart</td>
<td align="right">55</td>
<td align="right">11</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="even">
<td align="left">ENSE00001425120</td>
<td align="left">PABPC3</td>
<td align="left">chr13</td>
<td align="right">25096044</td>
<td align="right">25096048</td>
<td align="right">5</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_3</td>
<td align="right">2</td>
<td align="left">;cg02113641;cg16134678</td>
<td align="left">cluster_3</td>
<td align="right">3</td>
<td align="right">25096136</td>
<td align="right">25099254</td>
<td align="left">+</td>
<td align="left">upstream</td>
<td align="right">-90</td>
<td align="right">88</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="odd">
<td align="left">ENSE00001560063</td>
<td align="left">FTSJ1</td>
<td align="left">chrX</td>
<td align="right">48475919</td>
<td align="right">48476024</td>
<td align="right">106</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_8</td>
<td align="right">2</td>
<td align="left">;cg16248169;cg00473354</td>
<td align="left">cluster_8</td>
<td align="right">8</td>
<td align="right">48476021</td>
<td align="right">48476396</td>
<td align="left">+</td>
<td align="left">overlapStart</td>
<td align="right">-49</td>
<td align="right">3</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="even">
<td align="left">ENSE00001951259</td>
<td align="left">FAM50A</td>
<td align="left">chrX</td>
<td align="right">154443605</td>
<td align="right">154444373</td>
<td align="right">769</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_9</td>
<td align="right">2</td>
<td align="left">;cg13219667;cg27123903</td>
<td align="left">cluster_9</td>
<td align="right">9</td>
<td align="right">154444141</td>
<td align="right">154444346</td>
<td align="left">+</td>
<td align="left">includeFeature</td>
<td align="right">-152</td>
<td align="right">27</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="odd">
<td align="left">ENSE00002053470</td>
<td align="left">HOXC13-AS</td>
<td align="left">chr12</td>
<td align="right">53927517</td>
<td align="right">53927561</td>
<td align="right">45</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_2</td>
<td align="right">2</td>
<td align="left">;cg04651603;cg08190858</td>
<td align="left">cluster_2</td>
<td align="right">2</td>
<td align="right">53935328</td>
<td align="right">53936369</td>
<td align="left">-</td>
<td align="left">downstream</td>
<td align="right">8830</td>
<td align="right">7767</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="even">
<td align="left">ENSE00003133510</td>
<td align="left">PEG3</td>
<td align="left">chr19</td>
<td align="right">56840422</td>
<td align="right">56840773</td>
<td align="right">352</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_6</td>
<td align="right">2</td>
<td align="left">;cg14849423;cg13960339</td>
<td align="left">cluster_6</td>
<td align="right">6</td>
<td align="right">56840587</td>
<td align="right">56840661</td>
<td align="left">-</td>
<td align="left">includeFeature</td>
<td align="right">63</td>
<td align="right">112</td>
<td align="left">NearestLocation</td>
</tr>
<tr class="odd">
<td align="left">ENSE00003869597</td>
<td align="left">LINC03021</td>
<td align="left">chr8</td>
<td align="right">2728435</td>
<td align="right">2728684</td>
<td align="right">250</td>
<td align="left">*</td>
<td align="right">2</td>
<td align="left">cluster_7</td>
<td align="right">2</td>
<td align="left">;cg23099839;cg11779808</td>
<td align="left">cluster_7</td>
<td align="right">7</td>
<td align="right">2728330</td>
<td align="right">2728577</td>
<td align="left">-</td>
<td align="left">overlapStart</td>
<td align="right">17</td>
<td align="right">105</td>
<td align="left">NearestLocation</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">path</span>,recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ziqing Wang, Michael Cassidy, Danielle Clarkson-Townsend, Tamar Sofer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
